### PR TITLE
Provide intrinsics for wasm32

### PR DESCRIPTION
**Pull request by 197g**
Address all the stable load/store primitives that the standard library provides. Even though it remarks that many of these are not commonly required by giving safe alternatives, let's provide them anyways so that there is no need to trust the optimizer. Also, this way we get to use the align-1 cases even where it would be cumbersome with the provided alternatives. Also it's quite simple in the previous style.

---

Sadly, miri does not support any of these intrinsics. I found it hard to get it to work at all. It does not support the wasm32-wasip1 (entry point) ABI and on the `wasm32-unknown-unknown` target it only executed doctests, lib tests were silently ignored? Not sure what is going on. A simple way to run these tests is through `wasmtime` on `wasm32-wasip1` where binaries produced by cargo can be used as is and hence a `runner` configuration ensures `cargo` can still be utilized directly.

### CODE REVIEW

File: .github/workflows/rust-ci.yml [hunk 1]
```diff
@@ -102,6 +102,51 @@ jobs:
...
+          toolchain: ${{ matrix.toolchain }}
+          target: wasm32-wasip1
+          components: rust-docs
+      - run: >
+          curl -L https://github.com/bytecodealliance/wasmtime/releases/download/dev/wasmtime-dev-x86_64-linux.tar.xz | tar x --xz -f - --strip-components 1 "wasmtime-dev-x86_64-linux/wasmtime"
+      - name: "Enable simd128 target feature"
+        env:
+          RUSTFLAGS: "-Dwarnings -Ctarget-feature=+simd128"
+        run: |
+          echo "RUSTFLAGS=${{ env.RUSTFLAGS }}" >> $GITHUB_ENV
```
Comment (197g) [inline .github/workflows/rust-ci.yml:129]: @okaneco This write is necessary to preserve env variables across job steps. All other targets must be reviewed with this in mind, I don't think features are actually covered by tests. (Rust needs a good automated requirements traceability tool :smile:).

Comment (okaneco) [reply .github/workflows/rust-ci.yml:129]: For other targets, I don't think we currently exercise any target features outside of the ones that explicitly inline the RUSTFLAGS so we should be ok? But will have to keep that in mind.

The only thing that comes to mind is the `+avx` demo I tried to add. The non-temporal doctest is maybe better off as an `is_x86_feature_detected!()` than a `cfg!` target feature bool.

File: src/wasm32.rs [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+        [i16; 4],
+        [u32; 2],
+        [i32; 2],
+        [f32; 2],
+        [u64; 1],
+        [i64; 1],
+        [f64; 1],
+        f64,
+        u64,
+        i64,
```
Comment (okaneco) [inline src/wasm32.rs:192]: Consistency nit, but you did `uN, iN, fN` order for every other macro.

File: src/wasm32.rs [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+mod tests {
+    use core::arch::wasm32::{self as arch, v128};
+
+    fn assert_v128_bytes<const N: usize>(val: v128, data: &[[u8; N]]) {
+        assert_eq!(
+            unsafe { core::mem::transmute::<_, [u8; 16]>(val) },
+            unsafe { core::slice::from_raw_parts(data.as_ptr() as *const u8, data.len() * N) }
+        );
+
+        // assert_eq!(arch::to_bytes(val), <[u8; N]>::flat(data))
```
Comment (okaneco) [inline src/wasm32.rs:268]: Unused?

File: src/wasm32.rs [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+    }
+
+    #[test]
+    #[cfg_attr(not(target_feature = "simd128"), ignore)]
```
Comment (okaneco) [inline src/wasm32.rs:272]: `cfg_attr` because it's not a baseline included feature, right?

Unlike the other arches

**Review by okaneco — COMMENTED** [review]
This looks mostly straightforward as far as the impl goes.

I like the `impl_N_bits_traits` macro for arrays and primitives.

File: src/wasm32.rs [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+    pub fn test_i16x8_load_extend_u8x8() {
+        #[target_feature(enable = "simd128")]
+        fn test(a: [u8; 8]) {
+            let v = super::u16x8_load_extend_u8x8(&a);
+            assert_v128_bytes(v, &a.map(|i| (i as i16).to_ne_bytes()));
+        }
+
+        let a = core::array::from_fn(|i| u8::MAX - i as u8);
```
Comment (okaneco) [inline src/wasm32.rs:293]: Could you add a comment stating the `uNxN` tests are checking for zero-extension?

We do `u8:MAX - i`, to fill the top bit and make sure it's not sign-extending.

File: src/wasm32.rs (hunk 4 above)
Comment (197g) [reply src/wasm32.rs:272]: Yes, for neither of wasm32-unknown and wasm32-wasip1, wasm32-wasip2. However, strangely, wasmtime does not seem to care. Also it is odd that we can execute functions with that additional target feature without an `unsafe` block. But miri loudly complains and does not support that target feature at all.

File: .github/workflows/rust-ci.yml (hunk 1 above)
Comment (197g) [reply .github/workflows/rust-ci.yml:129]: These blocks: 

https://github.com/okaneco/safe_unaligned_simd/blob/master/.github/workflows/rust-ci.yml#L45-L55

Based on my reading of the docs and the CI failures I had before, that middle step is not actually executing with modifying RUSTFLAGS and the reset does nothing.

Comment (197g) [reply .github/workflows/rust-ci.yml:129]: Reference: https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/use-variables#passing-values-between-steps-and-jobs-in-a-workflow

File: src/wasm32.rs (hunk 4 above)
Comment (197g) [reply src/wasm32.rs:272]: According to: https://github.com/rust-lang/rust/issues/74372#issuecomment-828696632

Actually linking a function that made use of `target_feature(enable = )` requires that feature to be enabled globally, due to wasm specifying that validation happens _always. Hence, the only way to reach a function and execute is if the feature was enabled by the runtime, consequently the safety precondition is already fulfilled by `_main` being executed at all.

Seems sketch tbh but alright.

File: .github/workflows/rust-ci.yml (hunk 1 above)
Comment (okaneco) [reply .github/workflows/rust-ci.yml:129]: Removing that part of the workflow in #13 
The proper way to do that can be added in another PR that actually needs it

**Comment by 197g** [comment]
No, I don't. Since writes still have mutable reference in their signature, this style only works for the read-APIs. wasm32 is very read-heavy in its intrinsics but x86 is not. If anything, wasm32 should be changed to the x86 style, it seemed less of a hassle for users but the asymmetry is a buzzkill. Still a draft.

**Comment by okaneco** [comment]
I'm fine with `#![allow(unused)]` since I assumed it would be needed for the 1Byte trait that x86 doesn't use so extra cfg would be needed otherwise, until those traits are exported publicly crate-wide (if ever).

Implementing `Is??BitsUnaligned` for the arch-specific types seems reasonable to me for x86 since there are larger register types you could be feeding into. But is there a use for having the `Is256` and eventually `Is512` for wasm `v128` when there's no equivalent `v256`/`v512`? That's making me lean towards thinking we don't need the `Is256Bits` v128 impl at this time.

**Review by okaneco — APPROVED** [review]
